src/common.rs

```diff
@@ -28,6 +29,36 @@ pub struct MessageHeader {
     pub flags: u8,
 }
 
+impl<T> From<T> for MessageHeader
+where
+    T: AsRef<[u8]>,
+{
+    fn from(buf: T) -> Self {
+        let buf = buf.as_ref();
+        debug_assert!(buf.len() >= MESSAGE_HEADER_LENGTH);
+        Self {
+            length: BigEndian::read_u32(&buf[..4]),
+            stream_id: BigEndian::read_u32(&buf[4..8]),
+            type_: buf[8],
+            flags: buf[9],
+        }
+    }
+}
+
+impl MessageHeader {
+    pub(crate) fn into_buf(self, mut buf: impl AsMut<[u8]>) {
```

Comment:
impl `Into` trait same way as `From` ?

Reply:
Added `impl From<MessageHeader> for Vec<u8>`.
It is implemented using `into_buf`. I feel the `into_buf` function is more general so I kept it.

Reply:
Why not `impl<T> From<T> for MessageHeader where T: AsMut<[u8]>` ?

Reply:
There are compilation errors.
```
error[E0210]: type parameter `T` must be covered by another type when it appears before the first local type (`common::MessageHeader`)
  --> src/common.rs:48:6
   |
48 | impl<T: AsMut<u8>> From<MessageHeader> for T {
   |      ^ type parameter `T` must be covered by another type when it appears before the first local type (`common::MessageHeader`)
   |
   = note: implementing a foreign trait is only possible if at least one of the types for which it is implemented is local, and no uncovered type parameters appear before that first local type
   = note: in this case, 'before' refers to the following order: `impl<..> ForeignTrait<T1, ..., Tn> for T0`, where `T0` is the first and `Tn` is the last
```

Reply:
Got it, thanks for checking!